src/expr/parser.rs

```diff
@@ -1,60 +1,123 @@
-named!(term <i64>,
+named!(term <Expr>,
   chain!(
     mut acc: factor  ~
-             many0!(
-               alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
-               )
-             ),
+      many0!(
+        alt!(
+          map!(preceded!(tag!("*"), term), |prod: Expr| {
+            match acc {
+              Expr::Literal(num) => {
+                acc = Expr::BinOp(Box::new(Expr::Literal(num)),
+                                  BinOp::Times, Box::new(prod));
+              },
+              Expr::BinOp(ref mut x, ref mut op, ref mut y) => {
+                // swap out each of acc's values
+                let inner_x = mem::replace(x, Box::new(prod));
+                let inner_op = mem::replace(op, BinOp::Times);
+                let inner_y = mem::replace(y, Box::new(Expr::Literal(0)));
+                mem::replace(y, Box::new(
+                        Expr::BinOp(inner_x, inner_op, inner_y)));
+              }
+            }
+          }) |
+          map!(preceded!(tag!("/"), term), |div: Expr| {
+            match acc {
+              Expr::Literal(num) => {
+                acc = Expr::BinOp(Box::new(Expr::Literal(num)), BinOp::Over,
+                                  Box::new(div));
+              },
+              Expr::BinOp(ref mut x, ref mut op, ref mut y) => {
+                let inner_x = mem::replace(x, Box::new(Expr::Literal(0)));
+                let inner_op = mem::replace(op, BinOp::Over);
+                let inner_y = mem::replace(y, Box::new(div));
+                mem::replace(x, Box::new(
+                        Expr::BinOp(inner_x, inner_op, inner_y)));
+              }
+            }
+          })
+        )
+      ),
     || { return acc }
   )
 );
 
-named!(pub expr <i64>,
+named!(pub expr <Expr>,
   chain!(
     mut acc: term  ~
-             many0!(
-               alt!(
-                 tap!(add: preceded!(tag!("+"), term) => acc = acc + add) |
-                 tap!(sub: preceded!(tag!("-"), term) => acc = acc - sub)
-               )
-             ),
+      many0!(
+        alt!(
+          map!(preceded!(tag!("+"), term), |add: Expr| {
+            match acc {
+              Expr::Literal(num) => {
+                acc = Expr::BinOp(Box::new(Expr::Literal(num)), BinOp::Plus,
+                                  Box::new(add));
+              },
+              Expr::BinOp(ref mut x, ref mut op, ref mut y) => {
+                // swap out each of acc's values
+                let inner_x = mem::replace(x, Box::new(add));
+                let inner_op = mem::replace(op, BinOp::Plus);
+                let inner_y = mem::replace(y, Box::new(Expr::Literal(0)));
+                mem::replace(y, Box::new(
+                        Expr::BinOp(inner_x, inner_op, inner_y)));
+              }
+            }
+          }) |
```

Comment:
So this ends up working. You match on `acc`, and in the `Literal` case you bind the `i64`  directly (which doesn't move `acc` because `i64: Copy`. In the `BinOp` case you bind all contents by mutable reference, you also avoid the move.

There is a more clean way though: you can make sure to refer to use `acc` only as a mutable reference. This also avoids the move, and shortens your code a bunch.